File: homework/hw-5/python/sql_test_2.py

import psycopg as pg
import config

def main():
    # connection info
    hst = config.HOST
    usr = config.USER
    pwd = config.PASSWORD
    dat = config.DATABASE

    # make a connection
    with pg.connect(host=hst, user=usr, password=pwd, dbname=dat) as cn:
        # get a category from the user
        user_input_1 = input('Please enter a pet type (dog, cat, etc): ')
        user_input_2 = input('Please enter a pet name keyword: ') 

        # result set
        with cn.cursor() as rs:
            # parameterized query (must use old python format strings)
            q = ('SELECT id, name, type '
                 'FROM pet '
                 'WHERE type = %s AND strpos(name, %s) != 0')

            # execute the query (note, second param must be tuple)
            rs.execute(q, (user_input_1, user_input_2))
        
            # display results
            for row in rs:
                print(f'{row[0]}, {row[1]}, {row[2]}') 
            
        
if __name__ == '__main__':
    main()

File: homework/hw-5/cpp/sql_test_2.cpp

#include <iostream>
#include <pqxx/pqxx>
#include "config.h"

using namespace std;
using namespace pqxx;

int main()
{
  try {
    // connection info
    string usr = config::USER;
    string pwd = config::PASSWORD;
    string hst = config::HOST;
    string dat = config::DATABASE;
    string url = "postgresql://" + usr + ":" + pwd + "@" + hst + "/" + dat;
    
    // create a connection
    connection cx{url};
    work tx(cx);
    
    // get a category from the user
    string user_input_1;
    string user_input_2;
    cout << "Please enter a pet type (dog, cat, etc): ";
    cin >> user_input_1;
    cout << "Please enter an appearance keyword: ";
    cin >> user_input_2;
    
    // create a (named) prepared statement
    string q =
      "SELECT id, name, type "
      "FROM pet "
      "WHERE type = $1 AND strpos(name, $2) != 0";
    cx.prepare("q1", q);          // can give the query a name
    result r{tx.exec_prepared("q1", user_input_1, user_input_2)};
    
    // print result
    for (auto row: r) {
      cout << row["id"].as<int>() << ", "
           << row["name"].c_str() << ", "
           << row["type"].c_str() << endl;
    }

    // not needed here, but good habit if updates
    tx.commit();
  }
  catch(sql_error const &e) {
    cerr << "SQL error: " << e.what() << endl;
    return 1;
  }
  catch(exception const &e) {
    cerr << "Error: " << e.what() << endl;
    return 1;
  }
}
